src/server.c

```diff
@@ -3252,10 +3252,16 @@ static void propagateNow(int dbid, robj **argv, int argc, int target) {
      * 2. The primary waits for the replica to replicate before exiting, see
      * `shutdown-timeout` in conf for more details. In this case, if the primary
      * has lost a slot during the SIGTERM waiting, it will delete all the keys
-     * and replicates DEL to its replica. */
+     * and replicates DEL to its replica.
+     *
+     * 3. The primary waits for the replica to replicate during a manual failover.
+     * In this case, if the primary has lost a slot during the pausing, it will
+     * delete all the keys and replicates DEL to its replica. */
     int is_pause_replica_action = isPausedActions(PAUSE_ACTION_REPLICA);
     int is_pause_shutdown_purpose = isPausedPurpose(PAUSE_DURING_SHUTDOWN);
-    serverAssert(!is_pause_replica_action || server.client_pause_in_transaction || is_pause_shutdown_purpose);
+    int is_pause_failover_purpose = isPausedPurpose(PAUSE_DURING_FAILOVER);
+    serverAssert(!is_pause_replica_action || server.client_pause_in_transaction || is_pause_shutdown_purpose ||
+                 is_pause_failover_purpose);
```

Comment:
Now, there is only one more purpose: Paused by the CLIENT PAUSE command.

I guess a slot can be lost even when paused by CLIENT PAUSE, right? It can happen for any pause purpose, so just delete the assert?

How do we handle active expire? Is it prevented during pause? If not, this too can replicate DEL...

Reply:
we will also pause the expire, all the pause call like WRITE_SET and ALL_SET contains it.
```
/* Actions pause types */
#define PAUSE_ACTION_CLIENT_WRITE (1 << 0)
#define PAUSE_ACTION_CLIENT_ALL (1 << 1) /* must be bigger than PAUSE_ACTION_CLIENT_WRITE */
#define PAUSE_ACTION_EXPIRE (1 << 2)
#define PAUSE_ACTION_EVICT (1 << 3)
#define PAUSE_ACTION_REPLICA (1 << 4) /* pause replica traffic */

/* common sets of actions to pause/unpause */
#define PAUSE_ACTIONS_CLIENT_WRITE_SET \
    (PAUSE_ACTION_CLIENT_WRITE | PAUSE_ACTION_EXPIRE | PAUSE_ACTION_EVICT | PAUSE_ACTION_REPLICA)
#define PAUSE_ACTIONS_CLIENT_ALL_SET \
    (PAUSE_ACTION_CLIENT_ALL | PAUSE_ACTION_EXPIRE | PAUSE_ACTION_EVICT | PAUSE_ACTION_REPLICA)
```

yes, the client pause can also trigger this, i also thought of this last night after  turned off my mac (it was the bedtime). I guess we need a new var just like client_pause_in_transaction to track the state.